gobblin-core/src/main/java/gobblin/writer/AvroHdfsTimePartitionedWithRecordCountsWriter.java

```diff
@@ -0,0 +1,92 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.writer;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Map.Entry;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.State;
+import gobblin.util.ForkOperatorUtils;
+import gobblin.util.HadoopUtils;
+import gobblin.util.WriterUtils;
+
+
+/**
+ * Implementation of {@link DataWriter} that writes data into a date-partitioned directory structure based on the value
+ * of a specific field in each Avro record, while also including in the name of each file the number of records
+ * contained within that file.
+ *
+ * See {@link AvroHdfsTimePartitionedWriter} for more details on complete output path. This DataWriter maintains
+ * the same behavior except for including the number of records in the file name immediately before the extension.
+ * Where AvroHdfsTimePartitionedWriter would output path/fileName.avro,
+ * AvroHdfsTimePartitionedWithRecordCountsWriter would output path/fileName.{recordCount}.avro, e.g.
+ * path/fileName.450.avro
+ *
+ * This Writer behaves exactly as the AvroHdfsTimePartitionedWriter, but renames the outputted Avro files after
+ * they are closed.
+ */
+public class AvroHdfsTimePartitionedWithRecordCountsWriter extends AvroHdfsTimePartitionedWriter {
+
+  private final String writerId;
+  private final WriterOutputFormat writerOutputFormat;
+  private final State properties;
+  private final int numBranches;
+  private final int branch;
+
+  public AvroHdfsTimePartitionedWithRecordCountsWriter(Destination destination, String writerId, Schema schema,
+      WriterOutputFormat writerOutputFormat, int numBranches, int branch) {
+    super(destination, writerId, schema, writerOutputFormat, numBranches, branch);
+
+    this.writerId = writerId;
+    this.writerOutputFormat = writerOutputFormat;
+    this.numBranches = numBranches;
+    this.branch = branch;
+    this.properties = destination.getProperties();
+  }
+
+  @Override
+  public void close() throws IOException {
+    super.close();
+
+    String fileName = WriterUtils.getWriterFileName(this.properties, this.numBranches, this.branch, this.writerId,
+        this.writerOutputFormat.getExtension());
+
+    Configuration conf = new Configuration();
+    URI fileSystemURI = URI.create(this.properties.getProp(ForkOperatorUtils
+            .getPropertyNameForBranch(ConfigurationKeys.WRITER_FILE_SYSTEM_URI, this.numBranches, this.branch),
+        ConfigurationKeys.LOCAL_FS_URI));
+    FileSystem fileSystem = FileSystem.get(fileSystemURI, conf);
```

Comment:
Actually regardless to keep the logic centralized I agree that having a getFs() would be better, I will do so.